api/dataset.go

```diff
@@ -2,38 +2,218 @@ package api
+			if _, ok := dimensionMap[dimension.Variable.Name][option.Label]; !ok {
+				return nil, errors.New("option mismatch")
+			}
+
+			option := getDatasetJsonResponseDimensionOption{
+				Href: dimensionMap[dimension.Variable.Name][option.Label].Links.Code.URL,
+				Id:   option.Label,
+			}
+
+			options = append(options, option)
+		}
+
+		dimensions = append(dimensions, getDatasetJsonResponseDimension{
+			DimensionName: dimension.Variable.Name,
+			Options:       options,
+		})
+	}
+
+	getDatasetJsonResponse := getDatasetJsonObservationsResponse{
+		Dimensions:        dimensions,
+		Observations:      query.Dataset.Table.Values,
+		TotalObservations: len(query.Dataset.Table.Values),
+	}
+
+	return &getDatasetJsonResponse, nil
+}
+
+func (api *API) getDatasetParams(ctx context.Context, r *http.Request) (*datasetParams, error) {
+	params := &datasetParams{
+		id:      chi.URLParam(r, "dataset_id"),
+		edition: chi.URLParam(r, "edition"),
+		version: chi.URLParam(r, "version"),
+		options: make(optionsMap),
+	}
+
+	if params.id == "" {
+		return nil, errors.New("invalid dataset id")
+	}
+
+	if params.edition == "" {
+		return nil, errors.New("invalid edition")
+	}
+
+	if params.version == "" {
+		return nil, errors.New("invalid version")
+	}
+
+	datasetItem, err := api.datasets.GetDatasetCurrentAndNext(ctx, "", "", "", params.id)
+
+	if err != nil {
+		return nil, err
+	}
+
+	params.basedOn = datasetItem.IsBasedOn.ID
+
+	if datasetItem.DatasetDetails.Type != "cantabular_flexible_table" {
+		return nil, errors.New("invalid dataset type")
+	}
+
+	dimensions, err := api.datasets.GetVersionDimensions(ctx, "", "", "", params.id, params.edition, params.version)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if dimensions.Items.Len() == 0 {
+		return nil, errors.New("invalid dimensions length")
+	}
+
+	for _, dimension := range dimensions.Items {
+		options, err := api.datasets.GetOptionsInBatches(ctx, "", "", "", params.id, params.edition, params.version, dimension.Name, api.cfg.DatasetOptionsBatchSize, api.cfg.DatasetOptionsWorkers)
+
+		if err != nil {
+			return nil, err
+		}
+
+		params.options[dimension.Links.CodeList.ID] = make(map[string]dataset.Option)
+
+		for _, option := range options.Items {
+			params.options[dimension.Links.CodeList.ID][option.Label] = option
+		}
+
+		params.unsortedDimensions = append(params.unsortedDimensions, dimension.Links.CodeList.ID)
+	}
+
+	return params, nil
+}
+
+func (api *API) getGeographyTypes(ctx context.Context, datasetId string) ([]string, error) {
+	var geoDimensions []string
+
+	res, err := api.ctblr.GetGeographyDimensions(ctx, datasetId)
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, dimension := range res.Dataset.RuleBase.IsSourceOf.Edges {
+		geoDimensions = append(geoDimensions, dimension.Node.Name)
+	}
+
+	return geoDimensions, nil
+}
+
+func (api *API) sortGeography(geoDimensions []string, dimensions []string) ([]string, bool) {
+	foundGeography := false
+	sortedDimensions := []string{}
```

Comment:
It's bad practice to initialise slices before they're used unless necessary to avoid pre-emptive memory allocation. You should initialise the slices using `var` where possible.